From a315cec161935594cff2ca358281938f1e1c2fca Mon Sep 17 00:00:00 2001
Message-Id: <a315cec161935594cff2ca358281938f1e1c2fca.1421390813.git.chang-joon.lee@intel.com>
In-Reply-To: <55165e324aa18c65d2b754de3f9d891b8a8477e9.1421390813.git.chang-joon.lee@intel.com>
References: <55165e324aa18c65d2b754de3f9d891b8a8477e9.1421390813.git.chang-joon.lee@intel.com>
From: Vidya Srinivas <vidya.srinivas@intel.com>
Date: Fri, 28 Nov 2014 10:12:32 +0530
Subject: [PATCH 5/6] SQUASHME! [VPG]: drm/i915: Fix corruption related to
 protected video content

SQUASHME! - This patch should be squashed into the following exising patch:
	Author: Arun R Murthy <arun.r.murthy@intel.com>
	Change-Id: Id9fb3280aa59680a27536658dcaa335504a02e31
	REVERTME [VPG]: drm/i915: Add atomicity to i915.

This patch fixes corruption seen in protected video content
due to RRB2 programming as per atomicity.

Change-Id: I90132498031c94513f0721e5bf5e82b2df238f22
Signed-off-by: Arun R Murthy <arun.r.murthy@intel.com>
Signed-off-by: Jason Barstow <jason.barstow@intel.com>
Signed-off-by: Vidya Srinivas <vidya.srinivas@intel.com>
---
 drivers/gpu/drm/i915/intel_display.c |   12 ++++--------
 drivers/gpu/drm/i915/intel_sprite.c  |   20 +++++++++-----------
 2 files changed, 13 insertions(+), 19 deletions(-)

diff --git a/drivers/gpu/drm/i915/intel_display.c b/drivers/gpu/drm/i915/intel_display.c
index 50eeaf0..7d4947e 100644
--- a/drivers/gpu/drm/i915/intel_display.c
+++ b/drivers/gpu/drm/i915/intel_display.c
@@ -10542,7 +10542,7 @@ static void i915_commit(struct drm_i915_private *dev_priv,
 	struct drm_crtc *crtc = dev_priv->pipe_to_crtc_mapping[pipe];
 	struct intel_crtc *intel_crtc = to_intel_crtc(crtc);
 	struct intel_disp_reg *reg;
-	int plane;
+	int plane = 0;
 
 	if (type == SPRITE_PLANE) {
 		reg = &intel_plane->reg;
@@ -10583,7 +10583,7 @@ static void i915_commit(struct drm_i915_private *dev_priv,
 		I915_WRITE(SPLINOFF(pipe, plane), reg->linoff);
 		I915_WRITE(SPSIZE(pipe, plane),	reg->size);
 		I915_WRITE_BITS(SPCNTR(pipe, plane), reg->cntr, 0xFFFFFFF8);
-		I915_MODIFY_DISPBASE(SPSURF(pipe, plane), reg->surf);
+		I915_WRITE(SPSURF(pipe, plane), reg->surf);
 		if (intel_plane->pri_update) {
 			I915_WRITE(DSPCNTR(pipe), reg->dspcntr);
 			I915_MODIFY_DISPBASE(DSPSURF(pipe),
@@ -10599,8 +10599,6 @@ static void i915_commit(struct drm_i915_private *dev_priv,
 		I915_MODIFY_DISPBASE(DSPSURF(pipe), reg->surf);
 		POSTING_READ(DSPCNTR(pipe));
 	}
-
-	/* Reset the register */
 	reg->surf = 0;
 }
 
@@ -10687,12 +10685,10 @@ int intel_set_disp_plane_update(struct drm_mode_set_display *disp,
 
 		/* pass rrb2 information */
 		if (disp->plane[i].update_flag &
-			DRM_MODE_SET_DISPLAY_PLANE_UPDATE_RRB2) {
-			intel_plane->flags |=
-				DRM_MODE_SET_DISPLAY_PLANE_UPDATE_RRB2;
+			DRM_MODE_SET_DISPLAY_PLANE_UPDATE_RRB2)
 			intel_plane->rrb2_enable =
 				disp->plane[i].rrb2_enable;
-		}
+
 		if (disp->plane[i].update_flag &
 			DRM_MODE_SET_DISPLAY_PLANE_UPDATE_ALPHA) {
 			intel_plane->flags |=
diff --git a/drivers/gpu/drm/i915/intel_sprite.c b/drivers/gpu/drm/i915/intel_sprite.c
index 9e848d8..05b2dd0 100644
--- a/drivers/gpu/drm/i915/intel_sprite.c
+++ b/drivers/gpu/drm/i915/intel_sprite.c
@@ -429,17 +429,6 @@ vlv_update_plane(struct drm_plane *dplane, struct drm_crtc *crtc,
 		intel_plane->flags &= ~DRM_MODE_SET_DISPLAY_PLANE_UPDATE_ALPHA;
 	}
 
-	/* calculate the plane rrb2 */
-	if (intel_plane->flags & DRM_MODE_SET_DISPLAY_PLANE_UPDATE_RRB2) {
-		if (intel_plane->rrb2_enable)
-			intel_plane->reg.surf |=
-				PLANE_RESERVED_REG_BIT_2_ENABLE;
-		else
-			intel_plane->reg.surf &=
-				~PLANE_RESERVED_REG_BIT_2_ENABLE;
-		intel_plane->flags &= ~DRM_MODE_SET_DISPLAY_PLANE_UPDATE_RRB2;
-	}
-
 	switch (fb->pixel_format) {
 	case DRM_FORMAT_YUYV:
 		sprctl |= SP_FORMAT_YUV422 | SP_YUV_ORDER_YUYV;
@@ -683,7 +672,15 @@ vlv_update_plane(struct drm_plane *dplane, struct drm_crtc *crtc,
 	if ((sprite_ddl & mask) != (I915_READ(VLV_DDL(pipe)) & mask))
 		I915_WRITE_BITS(VLV_DDL(pipe), 0x00, mask);
 
+	intel_plane->reg.surf = I915_READ(SPSURF(pipe, plane));
+
+	if (intel_plane->rrb2_enable)
+		intel_plane->reg.surf |= PLANE_RESERVED_REG_BIT_2_ENABLE;
+	else
+		intel_plane->reg.surf &= ~PLANE_RESERVED_REG_BIT_2_ENABLE;
+
 	intel_plane->reg.cntr = sprctl;
+	intel_plane->reg.surf &= ~DISP_BASEADDR_MASK;
 	intel_plane->reg.surf |= i915_gem_obj_ggtt_offset(obj) + sprsurf_offset;
 	if (!dev_priv->atomic_update) {
 		I915_WRITE(SPCNTR(pipe, plane), sprctl);
@@ -2101,6 +2098,7 @@ intel_plane_init(struct drm_device *dev, enum pipe pipe, int plane)
 	intel_plane->pipe = pipe;
 	intel_plane->plane = plane;
 	intel_plane->rotate180 = false;
+	intel_plane->rrb2_enable = 0;
 	intel_plane->last_plane_state = INTEL_PLANE_STATE_DISABLED;
 	possible_crtcs = (1 << pipe);
 	ret = drm_plane_init(dev, &intel_plane->base, possible_crtcs,
-- 
1.7.9.5

